src/System.Windows.Forms/System/Windows/Forms/Dialogs/MessageBox.cs

```diff
@@ -13,12 +14,222 @@ public class MessageBox
+    {
+        BOOL currentValue;
+        HRESULT result = PInvoke.DwmGetWindowAttribute(
+          hWnd,
+           DWMWINDOWATTRIBUTE.DWMWA_USE_IMMERSIVE_DARK_MODE,
+          &currentValue,
+          (uint)sizeof(BOOL));
+        if (result.Succeeded)
+        {
+            if (currentValue == allow)
+            {
+                // no need for using DwmSetWindowAttribute
+                return;
+            }
+
+            PInvoke.DwmSetWindowAttribute(
+            hWnd,
+            DWMWINDOWATTRIBUTE.DWMWA_USE_IMMERSIVE_DARK_MODE,
+            &allow,
+            (uint)sizeof(BOOL)).AssertSuccess();
+        }
+    }
+
+    private static unsafe LRESULT DlgProc(IntPtr hWnd, int msg, IntPtr wparam, IntPtr lparam)
+    {
+        s_hWndInternal = (HWND)hWnd;
+        switch ((uint)msg)
+        {
+            case PInvokeCore.WM_INITDIALOG:
+#pragma warning disable WFO5001 // Type is for evaluation purposes only and is subject to change or removal in future updates.
+                AllowDarkNonClientArea(s_hWndInternal, Application.IsDarkModeEnabled);
+#pragma warning restore WFO5001 // Type is for evaluation purposes only and is subject to change or removal in future updates.
+                PInvokeCore.EnumChildWindows(s_hWndInternal, s_childWindowsCallback);
+                return PInvokeCore.CallWindowProc((void*)s_priorDlgProc, s_hWndInternal, (uint)msg, (nuint)wparam, lparam);
+            case PInvokeCore.WM_PAINT:
+                return OnWmPaint(s_hWndInternal, wparam);
+            case PInvokeCore.WM_CTLCOLORBTN:
+            case PInvokeCore.WM_CTLCOLORDLG:
+            case PInvokeCore.WM_CTLCOLORSTATIC:
+                return OnWmCtlColor((uint)msg, wparam);
+            default:
+                return PInvokeCore.CallWindowProc((void*)s_priorDlgProc, s_hWndInternal, (uint)msg, (nuint)wparam, lparam);
+        }
+    }
+
+    private static unsafe void InstallHook()
+    {
+        lock (s_lock)
+        {
+            if (s_messageBoxHook != 0)
+            {
+                return;
+            }
+
+            // see https://learn.microsoft.com/windows/win32/api/winuser/nf-winuser-setwindowshookexw
+            // Installs a hook procedure that monitors messages before the system sends them to the destination.
+
+            // Handling messages from WH_CALLWNDPROC or WH_CALLWNDPROCRET does not give good results so we chose to use WH_CALLWNDPROC and support subclassing by using Setwindowlong.
+            s_messageBoxHook = PInvoke.SetWindowsHookEx(
+                WINDOWS_HOOK_ID.WH_CALLWNDPROC,
+                (delegate* unmanaged[Stdcall]<int, WPARAM, LPARAM, LRESULT>)s_hookPointer,
+                HINSTANCE.Null,
+                PInvokeCore.GetCurrentThreadId());
+
+            s_isMessageBoxHooked = s_messageBoxHook != 0;
+
+            Debug.Assert(s_isMessageBoxHooked, "Failed to install MessageBox hook.");
+        }
+    }
+
+    private static void UninstallHook()
+    {
+        lock (s_lock)
```

Comment:
As said below add in this:
```suggestion
        if (!Application.IsDarkModeEnabled)
        {
            return;
        }

        lock (s_lock)
```